File: test/backend_test/src/com/jasonzqshen/familyAccountingBackendTest/utils/TransactionDataChecker.java
package com.jasonzqshen.familyAccountingBackendTest.utils;

import static org.junit.Assert.*;

import java.util.Calendar;

import com.jasonzqshen.familyaccounting.core.CoreDriver;
import com.jasonzqshen.familyaccounting.core.transaction.HeadEntity;
import com.jasonzqshen.familyaccounting.core.transaction.ItemEntity;
import com.jasonzqshen.familyaccounting.core.transaction.MonthIdentity;
import com.jasonzqshen.familyaccounting.core.transaction.MonthLedger;
import com.jasonzqshen.familyaccounting.core.transaction.TransactionDataManagement;
import com.jasonzqshen.familyaccounting.core.utils.AccountType;
import com.jasonzqshen.familyaccounting.core.utils.CreditDebitIndicator;
import com.jasonzqshen.familyaccounting.core.utils.DocumentType;

public class TransactionDataChecker {
	/**
	 * check transaction data
	 * 
	 * @param coreDriver
	 */
	public static void checkTransactionData(CoreDriver coreDriver) {
		TransactionDataManagement transManagement = coreDriver
				.getTransDataManagement();
		MonthIdentity[] monthIds = transManagement.getAllMonthIds();
		assertEquals(2, monthIds.length);

		MonthLedger ledger07 = transManagement.getLedger(monthIds[0]);
		MonthLedger ledger08 = transManagement.getLedger(monthIds[1]);
		assertEquals(true, ledger07.isClosed());
		assertEquals(false, ledger08.isClosed());
		assertTrue(transManagement.getCurrentLedger() == ledger08);

		checkLedger2012_07(ledger07.getEntities());
		checkLedger2012_08(ledger08.getEntities());
	}

	/**
	 * check ledger 2012_07
	 * 
	 * @param collection
	 */
	public static void checkLedger2012_07(HeadEntity[] docs) {
		assertEquals(3, docs.length);

		// check values
		for (int i = 0; i < docs.length; ++i) {
			HeadEntity head = docs[i];
			// posting date
			Calendar cal = Calendar.getInstance();
			cal.setTime(head.getPostingDate());
			assertEquals(2012, cal.get(Calendar.YEAR));
			assertEquals(7, cal.get(Calendar.MONTH) + 1);
			assertEquals(2, cal.get(Calendar.DATE));

			// document type
			if (i == 0) {
				assertEquals(DocumentType.VENDOR_INVOICE,
						head.getDocumentType());
			} else if (i == 1) {
				assertEquals(DocumentType.CUSTOMER_INVOICE,
						head.getDocumentType());
			} else {
				assertEquals(DocumentType.GL, head.getDocumentType());
			}

			// is reversed
			assertEquals(false, head.IsReversed());
			assertEquals(null, head.getReference());

			if (i == 2) {
				checkClosingDoc(head);
			} else {
				// text
				assertEquals(TestUtilities.TEST_DESCP, head.getDocText());

				// items
				ItemEntity[] items = head.getItems();
				assertEquals(2, items.length);
				if (i == 0) {
					checkCostItem(items[0]);
					checkVendorItem(items[1]);
				} else {
					checkProfitItem(items[0]);
					checkCustomerItem(items[1]);
				}
			}

		}
	}

	/**
	 * check ledger 2012_07
	 * 
	 * @param collection
	 */
	public static void checkLedger2012_08(HeadEntity[] docs) {
		assertEquals(2, docs.length);

		// check values
		for (int i = 0; i < docs.length; ++i) {
			HeadEntity head = docs[i];
			// posting date
			Calendar cal = Calendar.getInstance();
			cal.setTime(head.getPostingDate());
			assertEquals(2012, cal.get(Calendar.YEAR));
			assertEquals(8, cal.get(Calendar.MONTH) + 1);
			assertEquals(2, cal.get(Calendar.DATE));

			// document type
			assertEquals(DocumentType.VENDOR_INVOICE, head.getDocumentType());

			// is reversed
			assertEquals(true, head.IsReversed());
			assertEquals(docs[1].getDocIdentity(), head.getReference());

			if (i == 2) {
				checkClosingDoc(head);
			} else {
				// text
				assertEquals(TestUtilities.TEST_DESCP, head.getDocText());

				// items
				ItemEntity[] items = head.getItems();
				assertEquals(2, items.length);
				if (i == 0) {
					checkCostItem(items[0]);
					checkVendorItem(items[1]);
				} else {
					checkCostItemRev(items[0]);
					checkVendorItemRev(items[1]);
				}
			}

		}
	}

	/**
	 * check closing document
	 * 
	 * @param closingDoc
	 */
	private static void checkClosingDoc(HeadEntity closingDoc) {
		// text
		assertEquals(MonthLedger.CLOSING_DOC_TAG, closingDoc.getDocText());
		// items
		ItemEntity[] items = closingDoc.getItems();
		assertEquals(3, items.length);

		checkProfitItemRev(items[0]);
		ItemEntity item = items[1];
		assertEquals(TestUtilities.GL_ACCOUNT_COST, item.getGLAccount()
				.toString());
		assertEquals(null, item.getVendor());
		assertEquals(null, item.getCustomer());
		assertEquals(AccountType.GL_ACCOUNT, item.getAccountType());
		assertEquals(TestUtilities.TEST_AMOUNT1, item.getAmount().toString());
		assertEquals(CreditDebitIndicator.CREDIT, item.getCDIndicator());
		
		
		item = items[2];
		assertEquals(TestUtilities.GL_ACCOUNT_ENQUITY, item.getGLAccount()
				.toString());
		assertEquals(null, item.getVendor());
		assertEquals(null, item.getCustomer());
		assertEquals(AccountType.GL_ACCOUNT, item.getAccountType());
		assertEquals(TestUtilities.TEST_AMOUNT3, item.getAmount().toString());
		assertEquals(CreditDebitIndicator.CREDIT, item.getCDIndicator());
	}

	/**
	 * check cost item
	 * 
	 * @param costItem
	 */
	private static void checkCostItem(ItemEntity item) {
		assertEquals(TestUtilities.GL_ACCOUNT_COST, item.getGLAccount()
				.toString());
		assertEquals(null, item.getVendor());
		assertEquals(null, item.getCustomer());
		assertEquals(AccountType.GL_ACCOUNT, item.getAccountType());
		assertEquals(TestUtilities.TEST_AMOUNT1, item.getAmount().toString());
		assertEquals(CreditDebitIndicator.DEBIT, item.getCDIndicator());

		assertEquals(TestUtilities.BUSINESS_AREA, item.getBusinessArea()
				.toString());
	}

	/**
	 * check vendor item
	 * 
	 * @param costItem
	 */
	private static void checkVendorItem(ItemEntity item) {
		assertEquals(TestUtilities.GL_ACCOUNT2, item.getGLAccount().toString());
		assertEquals(TestUtilities.VENDOR, item.getVendor().toString());
		assertEquals(null, item.getCustomer());
		assertEquals(AccountType.VENDOR, item.getAccountType());
		assertEquals(TestUtilities.TEST_AMOUNT1, item.getAmount().toString());
		assertEquals(CreditDebitIndicator.CREDIT, item.getCDIndicator());

		assertEquals(null, item.getBusinessArea());
	}

	/**
	 * check cost item
	 * 
	 * @param costItem
	 */
	private static void checkCostItemRev(ItemEntity item) {
		assertEquals(TestUtilities.GL_ACCOUNT_COST, item.getGLAccount()
				.toString());
		assertEquals(null, item.getVendor());
		assertEquals(null, item.getCustomer());
		assertEquals(AccountType.GL_ACCOUNT, item.getAccountType());
		assertEquals(TestUtilities.TEST_AMOUNT1, item.getAmount().toString());
		assertEquals(CreditDebitIndicator.CREDIT, item.getCDIndicator());

		assertEquals(TestUtilities.BUSINESS_AREA, item.getBusinessArea()
				.toString());
	}

	/**
	 * check vendor item
	 * 
	 * @param costItem
	 */
	private static void checkVendorItemRev(ItemEntity item) {
		assertEquals(TestUtilities.GL_ACCOUNT2, item.getGLAccount().toString());
		assertEquals(TestUtilities.VENDOR, item.getVendor().toString());
		assertEquals(null, item.getCustomer());
		assertEquals(AccountType.VENDOR, item.getAccountType());
		assertEquals(TestUtilities.TEST_AMOUNT1, item.getAmount().toString());
		assertEquals(CreditDebitIndicator.DEBIT, item.getCDIndicator());

		assertEquals(null, item.getBusinessArea());
	}

	/**
	 * check profit item
	 * 
	 * @param costItem
	 */
	private static void checkProfitItem(ItemEntity item) {
		assertEquals(TestUtilities.GL_ACCOUNT_PROFIT, item.getGLAccount()
				.toString());
		assertEquals(null, item.getVendor());
		assertEquals(null, item.getCustomer());
		assertEquals(AccountType.GL_ACCOUNT, item.getAccountType());
		assertEquals(TestUtilities.TEST_AMOUNT2, item.getAmount().toString());
		assertEquals(CreditDebitIndicator.CREDIT, item.getCDIndicator());

		assertEquals(null, item.getBusinessArea());
	}

	/**
	 * check profit item
	 * 
	 * @param costItem
	 */
	private static void checkProfitItemRev(ItemEntity item) {
		assertEquals(TestUtilities.GL_ACCOUNT_PROFIT, item.getGLAccount()
				.toString());
		assertEquals(null, item.getVendor());
		assertEquals(null, item.getCustomer());
		assertEquals(AccountType.GL_ACCOUNT, item.getAccountType());
		assertEquals(TestUtilities.TEST_AMOUNT2, item.getAmount().toString());
		assertEquals(CreditDebitIndicator.DEBIT, item.getCDIndicator());

		assertEquals(null, item.getBusinessArea());
	}

	/**
	 * check customer item
	 * 
	 * @param costItem
	 */
	private static void checkCustomerItem(ItemEntity item) {
		assertEquals(TestUtilities.GL_ACCOUNT2, item.getGLAccount().toString());
		assertEquals(TestUtilities.CUSTOMER, item.getCustomer().toString());
		assertEquals(null, item.getVendor());
		assertEquals(AccountType.CUSTOMER, item.getAccountType());
		assertEquals(TestUtilities.TEST_AMOUNT2, item.getAmount().toString());
		assertEquals(CreditDebitIndicator.DEBIT, item.getCDIndicator());

		assertEquals(null, item.getBusinessArea());
	}
}

File: trunk/proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingUnitTest/Utils/TransactionDataChecker.cs
using Microsoft.VisualStudio.TestPlatform.UnitTestFramework;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WinStore_FamilyAccountingCore;
using WinStore_FamilyAccountingCore.MasterData;
using WinStore_FamilyAccountingCore.Transaction;
using WinStore_FamilyAccountingCore.Utilities;

namespace WinStore_FamilyAccountingUnitTest.Utils
{
    public class TransactionDataChecker
    {
        /// <summary>
        /// check transaction data
        /// </summary>
        /// <param name="coreDriver"></param>
        public static void CheckTransactionData(CoreDriver coreDriver)
        {
            TransactionDataManagement transManagement = coreDriver.TransMgmt;
            MonthIdentity[] monthIds = coreDriver.MonthIds;
            // Assert.AreEqual(2, monthIds.length);

            MonthLedger ledger07 = transManagement.GetLedger(monthIds[0]);
            MonthLedger ledger08 = transManagement.GetLedger(monthIds[1]);

            CheckLedger2012_07(ledger07.Entities);
            CheckLedger2012_08(ledger08.Entities);
        }

        /// <summary>
        /// check ledger
        /// </summary>
        /// <param name="docs"></param>
        public static void CheckLedger2012_07(List<HeadEntity> docs)
        {
            Assert.AreEqual(3, docs.Count);

            checkVendorDoc(docs[0]);
            checkCustomerDoc(docs[1]);
            checkGLDoc(docs[2]);
        }

        /// <summary>
        /// check vendor document
        /// </summary>
        /// <param name="vendorDoc"></param>
        private static void checkVendorDoc(HeadEntity vendorDoc)
        {
            // check document number
            Assert.AreEqual(TestData.DOC_NUM1, vendorDoc.DocNumber
                    .ToString());
            // check posting date
            Assert.AreEqual(TestData.DATE_2012_07, vendorDoc.PstDate.ToString(MasterDataUtils.DATE_FORMAT));
            MonthIdentity monthId = vendorDoc.MonthID;
            Assert.AreEqual(TestData.YEAR, monthId.FiscalYear);
            Assert.AreEqual(TestData.MONTH_07, monthId.FiscalMonth);
            // check document type
            Assert.AreEqual(DocumentType.VENDOR_INVOICE, vendorDoc.DocType);
            // check text
            Assert.AreEqual(TestData.TEXT_VENDOR_DOC, vendorDoc.DocText);
            // check is reversed
            Assert.AreEqual(false, vendorDoc.IsReversed);

            Assert.AreEqual(2, vendorDoc.ItemCount);
            List<ItemEntity> items = vendorDoc.Items;
            // check the vendor item
            Assert.AreEqual(0, items[0].LineNum);
            Assert.AreEqual(AccountType.VENDOR, items[0].AccType);
            Assert.AreEqual(CreditDebitIndicator.CREDIT, items[0].CdIndicator);
            Assert.AreEqual(TestData.AMOUNT_VENDOR, items[0].Amount);
            Assert.AreEqual(null, items[0].Customer);
            Assert.AreEqual(TestData.VENDOR_BUS, items[0].Vendor.ToString());
            Assert.AreEqual(TestData.GL_ACCOUNT_CASH, items[0].GLAccount
                    .ToString());
            Assert.AreEqual(null, items[0].BusinessArea);

            // check cost item
            Assert.AreEqual(1, items[1].LineNum);
            Assert.AreEqual(AccountType.GL_ACCOUNT, items[1].AccType);
            Assert.AreEqual(CreditDebitIndicator.DEBIT, items[1].CdIndicator);
            Assert.AreEqual(TestData.AMOUNT_VENDOR, items[1].Amount);
            Assert.AreEqual(null, items[1].Customer);
            Assert.AreEqual(null, items[1].Vendor);
            Assert.AreEqual(TestData.GL_ACCOUNT_COST, items[1].GLAccount
                    .ToString());
            Assert.AreEqual(TestData.BUSINESS_AREA_WORK, items[1].BusinessArea
                    .ToString());
        }

        /// <summary>
        /// check customer document
        /// </summary>
        /// <param name="customerDoc"></param>
        private static void checkCustomerDoc(HeadEntity customerDoc)
        {
            // check document number
            Assert.AreEqual(TestData.DOC_NUM2, customerDoc.DocNumber
                    .ToString());
            // check posting date
            Assert.AreEqual(TestData.DATE_2012_07, customerDoc.PstDate.ToString(
                MasterDataUtils.DATE_FORMAT));
            MonthIdentity monthId = customerDoc.MonthID;
            Assert.AreEqual(TestData.YEAR, monthId.FiscalYear);
            Assert.AreEqual(TestData.MONTH_07, monthId.FiscalMonth);
            // check document type
            Assert.AreEqual(DocumentType.CUSTOMER_INVOICE,
                    customerDoc.DocType);
            // check text
            Assert.AreEqual(TestData.TEXT_CUSTOMER_DOC, customerDoc.DocText);
            // check is reversed
            Assert.AreEqual(false, customerDoc.IsReversed);

            Assert.AreEqual(2, customerDoc.ItemCount);
            List<ItemEntity> items = customerDoc.Items;
            // check revenue item
            Assert.AreEqual(0, items[0].LineNum);
            Assert.AreEqual(AccountType.GL_ACCOUNT, items[0].AccType);
            Assert.AreEqual(CreditDebitIndicator.CREDIT, items[0].CdIndicator);
            Assert.AreEqual(TestData.AMOUNT_CUSTOMER, items[0].Amount);
            Assert.AreEqual(null, items[0].Customer);
            Assert.AreEqual(null, items[0].Vendor);
            Assert.AreEqual(TestData.GL_ACCOUNT_REV, items[0].GLAccount
                    .ToString());
            Assert.AreEqual(null, items[0].BusinessArea);

            // check customer item
            Assert.AreEqual(1, items[1].LineNum);
            Assert.AreEqual(AccountType.CUSTOMER, items[1].AccType);
            Assert.AreEqual(CreditDebitIndicator.DEBIT, items[1].CdIndicator);
            Assert.AreEqual(TestData.AMOUNT_CUSTOMER, items[1].Amount);
            Assert.AreEqual(TestData.CUSTOMER1, items[1].Customer.ToString());
            Assert.AreEqual(null, items[1].Vendor);
            Assert.AreEqual(TestData.GL_ACCOUNT_BANK, items[1].GLAccount
                    .ToString());
            Assert.AreEqual(null, items[1].BusinessArea);
        }

        /// <summary>
        /// check gl doc
        /// </summary>
        /// <param name="glDoc"></param>
        private static void checkGLDoc(HeadEntity glDoc)
        {
            // check document number
            Assert.AreEqual(TestData.DOC_NUM3, glDoc.DocNumber.ToString());
            // check posting date
            Assert.AreEqual(TestData.DATE_2012_07, glDoc.PstDate.ToString(MasterDataUtils.DATE_FORMAT));
            MonthIdentity monthId = glDoc.MonthID;
            Assert.AreEqual(TestData.YEAR, monthId.FiscalYear);
            Assert.AreEqual(TestData.MONTH_07, monthId.FiscalMonth);
            // check document type
            Assert.AreEqual(DocumentType.GL, glDoc.DocType);
            // check text
            Assert.AreEqual(TestData.TEXT_GL_DOC, glDoc.DocText);
            // check is reversed
            Assert.AreEqual(false, glDoc.IsReversed);

            Assert.AreEqual(2, glDoc.ItemCount);
            List<ItemEntity> items = glDoc.Items;
            // check the source item
            Assert.AreEqual(0, items[0].LineNum);
            Assert.AreEqual(AccountType.GL_ACCOUNT, items[0].AccType);
            Assert.AreEqual(CreditDebitIndicator.CREDIT, items[0].CdIndicator);
            Assert.AreEqual(TestData.AMOUNT_GL, items[0].Amount);
            Assert.AreEqual(null, items[0].Customer);
            Assert.AreEqual(null, items[0].Vendor);
            Assert.AreEqual(TestData.GL_ACCOUNT_BANK, items[0].GLAccount
                    .ToString());
            Assert.AreEqual(null, items[0].BusinessArea);

            // check destination item
            Assert.AreEqual(1, items[1].LineNum);
            Assert.AreEqual(AccountType.GL_ACCOUNT, items[1].AccType);
            Assert.AreEqual(CreditDebitIndicator.DEBIT, items[1].CdIndicator);
            Assert.AreEqual(TestData.AMOUNT_GL, items[1].Amount);
            Assert.AreEqual(null, items[1].Customer);
            Assert.AreEqual(null, items[1].Vendor);
            Assert.AreEqual(TestData.GL_ACCOUNT_CASH, items[1].GLAccount
                    .ToString());
            Assert.AreEqual(null, items[1].BusinessArea);
        }

        /// <summary>
        /// check ledger 2012 07
        /// </summary>
        /// <param name="docs"></param>
        public static void CheckLedger2012_08(List<HeadEntity> docs)
        {
            Assert.AreEqual(1, docs.Count);
            checkVendorDoc_08(docs[0]);
        }


        /// <summary>
        /// check vendor doc in 08
        /// </summary>
        /// <param name="vendorDoc"></param>
        private static void checkVendorDoc_08(HeadEntity vendorDoc)
        {
            // check document number
            Assert.AreEqual(TestData.DOC_NUM1, vendorDoc.DocNumber
                    .ToString());
            // check posting date
            Assert.AreEqual(TestData.DATE_2012_08, vendorDoc.PstDate.ToString(MasterDataUtils.DATE_FORMAT));
            MonthIdentity monthId = vendorDoc.MonthID;
            Assert.AreEqual(TestData.YEAR, monthId.FiscalYear);
            Assert.AreEqual(TestData.MONTH_08, monthId.FiscalMonth);
            // check document type
            Assert.AreEqual(DocumentType.VENDOR_INVOICE, vendorDoc.DocType);
            // check text
            Assert.AreEqual(TestData.TEXT_VENDOR_DOC, vendorDoc.DocText);
            // check is reversed
            Assert.AreEqual(true, vendorDoc.IsReversed);

            Assert.AreEqual(2, vendorDoc.ItemCount);
            List<ItemEntity> items = vendorDoc.Items;
            // check the vendor item
            Assert.AreEqual(0, items[0].LineNum);
            Assert.AreEqual(AccountType.VENDOR, items[0].AccType);
            Assert.AreEqual(CreditDebitIndicator.CREDIT, items[0].CdIndicator);
            Assert.AreEqual(TestData.AMOUNT_VENDOR, items[0].Amount);
            Assert.AreEqual(null, items[0].Customer);
            Assert.AreEqual(TestData.VENDOR_BUS, items[0].Vendor.ToString());
            Assert.AreEqual(TestData.GL_ACCOUNT_CASH, items[0].GLAccount
                    .ToString());
            Assert.AreEqual(null, items[0].BusinessArea);

            // check cost item
            Assert.AreEqual(1, items[1].LineNum);
            Assert.AreEqual(AccountType.GL_ACCOUNT, items[1].AccType);
            Assert.AreEqual(CreditDebitIndicator.DEBIT, items[1].CdIndicator);
            Assert.AreEqual(TestData.AMOUNT_VENDOR, items[1].Amount);
            Assert.AreEqual(null, items[1].Customer);
            Assert.AreEqual(null, items[1].Vendor);
            Assert.AreEqual(TestData.GL_ACCOUNT_COST, items[1].GLAccount
                    .ToString());
            Assert.AreEqual(TestData.BUSINESS_AREA_WORK, items[1].BusinessArea
                    .ToString());
        }
    }
}
